tour_model_eval/confirmed_trips_eval_bins_clusters.py

```diff
@@ -20,54 +22,213 @@
+        # and 50% of total trips are valid) to analyze
+        if not valid_user(filter_trips,trips):
+            homo_score.append(NaN)
+            comp_score.append(NaN)
+            v_score.append(NaN)
+            continue
+        sim.bin_data()
+        if cutoff is None:
+            trip_index_ls = []
+            bins = sim.bins
+            for bin in bins:
+                for index in bin:
+                    trip_index_ls.append(index)
+            bin_trips = [filter_trips[num] for num in trip_index_ls]
+
+        elif cutoff:
+            sim.delete_bins()
+            bin_trips = sim.newdata
+            bins = sim.bins
+
+        bin_trips_df = pd.DataFrame(data=[trip["data"]["user_input"] for trip in bin_trips])
+
+        if sp2en:
+            bin_trips_df = bin_trips_df.replace(span_eng_dict)
+        elif cvt_pur_mo:
+            bin_trips_df = bin_trips_df.replace(span_eng_dict)
+            bin_trips_df = bin_trips_df.replace(map_pur_dict)
+            for a in range(len(bin_trips_df)):
+                if bin_trips_df.iloc[a]["replaced_mode"] == "same_mode":
+                    # to see which row will be converted
+                    logging.debug("The following rows will be changed: %s", bin_trips_df.iloc[a])
+                    bin_trips_df.iloc[a]["replaced_mode"] = bin_trips_df.iloc[a]['mode_confirm']
+
+        # turn all user_input into list without binning
+        bin_trips_user_input_ls = bin_trips_df.values.tolist()
+        # drop duplicate user_input
+        no_dup_df = bin_trips_df.drop_duplicates()
+        # turn non-duplicate user_input into list
+        no_dup_list = no_dup_df.values.tolist()
+
+        # collect labels_true based on user_input
+        labels_true = []
+        for trip in bin_trips_user_input_ls:
+            if trip in no_dup_list:
+                labels_true.append(no_dup_list.index(trip))
+
+        # collect labels_pred based on bins
+        labels_pred = []
+        for b in range(len(bins)):
+            for trip in bins[b]:
+                labels_pred.append(b)
+
+        # compare the trips order in bins and those in valid_trips using timestamp
+        bin_trips_ts = pd.DataFrame(data=[trip["data"]["start_ts"] for trip in bin_trips])
+        bin_ls = []
+        for bin in bins:
+            for index in bin:
+                bin_ls.append(index)
+        bins_ts = pd.DataFrame(data=[filter_trips[i]["data"]["start_ts"] for i in bin_ls])
+        # compare two data frames, the program will continue to score calculation if two data frames are the same
+        assert_frame_equal(bins_ts, bin_trips_ts)
+        homo = metrics.homogeneity_score(labels_true, labels_pred)
+        homo_score.append(float('%.3f' % homo))
+        comp = metrics.completeness_score(labels_true, labels_pred)
+        comp_score.append(float('%.3f' % comp))
+        v = metrics.v_measure_score(labels_true, labels_pred)
+        v_score.append(float('%.3f' % v))
```

Comment:
This should be pulled out into `compute_score`.
In general, you do not want to have any repeated code.
I think I have linked to DRY before 
https://en.wikipedia.org/wiki/Don%27t_repeat_yourself

Reply:
OK, I understand `valid_user_check` and `compute_score` now.
Can I modify the code after I figure out the solution for the second round clustering?

Reply:
@corinne-hcr I understand that you want to wait until the end to share the code. My concern is that then the code will be large and unwieldy and impossible to review. I would really like to review and merge intermediate code as we move along.

Can you please get this PR into a reviewable state ASAP so I can merge?